Reimplement <!-- webkit-test-runner: lang=something --> without using ProcessPoolConfiguration

litherum opened the pull request:
#### fb205fb01a0863848066ea78dbfb0e2bc67b44a1
<pre>
Reimplement &lt;!-- webkit-test-runner: lang=something --&gt; without using ProcessPoolConfiguration
<a href="https://bugs.webkit.org/show_bug.cgi?id=242584">https://bugs.webkit.org/show_bug.cgi?id=242584</a>

Reviewed by Wenson Hsieh.

<a href="https://bugs.webkit.org/show_bug.cgi?id=242581">https://bugs.webkit.org/show_bug.cgi?id=242581</a> is about how the GPU process needs to have the same system
language set as the web process that it&apos;s talking to. However, there is only one GPU process, regardless of how
many web process pools there are. Today, override languages (&lt;!-- webkit-test-runner: lang=something --&gt;) are
stored inside the ProcessPoolConfiguration, but there might be 2 process pools, so the GPU process wouldn&apos;t
know which override languages to use.

Luckily, these override languages are just used for testing, and our testing infrastructure doesn&apos;t need this
flexibility. This patch reimplements override languages, but stores them in a static NeverDestroyed in the
UI process, rather than in the ProcessPoolConfiguration. This way, the fix for
<a href="https://bugs.webkit.org/show_bug.cgi?id=242581">https://bugs.webkit.org/show_bug.cgi?id=242581</a> can have the GPU process creation code just read the static
override languages directly.

This is a testing-only change.

Covered by existing tests.

* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/API/APIProcessPoolConfiguration.h:
* Source/WebKit/UIProcess/API/C/WKContextConfigurationRef.cpp:
(WKContextConfigurationCopyOverrideLanguages): Deleted.
(WKContextConfigurationSetOverrideLanguages): Deleted.
* Source/WebKit/UIProcess/API/C/WKContextConfigurationRef.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivateForTesting.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewTesting.mm:
(+[WKWebView _setOverrideLanguagesForTesting:]):
* Source/WebKit/UIProcess/OverrideLanguages.cpp: Added.
(WebKit::overrideLanguagesStorage):
(WebKit::setOverrideLanguages):
(WebKit::overrideLanguages):
* Source/WebKit/UIProcess/OverrideLanguages.h: Copied from Source/WebKit/UIProcess/WebContextMenuProxy.h.
* Source/WebKit/UIProcess/WebContextMenuProxy.h:
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::setOverrideLanguages):
(WebKit::WebProcessPool::initializeNewWebProcess):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::getLaunchOptions):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::generateContextConfiguration const):
* Tools/WebKitTestRunner/cocoa/TestControllerCocoa.mm:
(WTR::TestController::cocoaResetStateToConsistentValues):

Canonical link: <a href="https://commits.webkit.org/252335@main">https://commits.webkit.org/252335@main</a>
</pre>

whsieh commented on Source/WebKit/UIProcess/WebProcessPool.cpp:
```diff
@@ -365,13 +366,13 @@ void WebProcessPool::setCustomWebContentServiceBundleIdentifier(const String& cu
 void WebProcessPool::setOverrideLanguages(Vector<String>&& languages)
 {
-    m_configuration->setOverrideLanguages(WTFMove(languages));
+    WebKit::setOverrideLanguages(WTFMove(languages));
```
Nit - can you omit the `WebKit::` here?

litherum replied:
Unfortunately not - that would make this function recursive. Do you think it's better to rename one of the two of them to avoid this problem?

whsieh replied:
Ah, right — that's fine, then.

whsieh commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -417,7 +418,7 @@ void WebProcessProxy::getLaunchOptions(ProcessLauncher::LaunchOptions& launchOpt
 
     LOG(Language, "WebProcessProxy is getting launch options.");
-    auto overrideLanguages = m_processPool->configuration().overrideLanguages();
+    auto overrideLanguages = WebKit::overrideLanguages();
```
(Ditto)

litherum replied:
Clang complains about this because it would be `auto foo = foo();` and it looks like the variable is used in its initializer. Do you think I should rename the variable to avoid this problem?

litherum committed:
Reimplement <!-- webkit-test-runner: lang=something --> without using ProcessPoolConfiguration
https://bugs.webkit.org/show_bug.cgi?id=242584

Reviewed by Wenson Hsieh.

https://bugs.webkit.org/show_bug.cgi?id=242581 is about how the GPU process needs to have the same system
language set as the web process that it's talking to. However, there is only one GPU process, regardless of how
many web process pools there are. Today, override languages (<!-- webkit-test-runner: lang=something -->) are
stored inside the ProcessPoolConfiguration, but there might be 2 process pools, so the GPU process wouldn't
know which override languages to use.

Luckily, these override languages are just used for testing, and our testing infrastructure doesn't need this
flexibility. This patch reimplements override languages, but stores them in a static NeverDestroyed in the
UI process, rather than in the ProcessPoolConfiguration. This way, the fix for
https://bugs.webkit.org/show_bug.cgi?id=242581 can have the GPU process creation code just read the static
override languages directly.

This is a testing-only change.

Covered by existing tests.

* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/API/APIProcessPoolConfiguration.h:
* Source/WebKit/UIProcess/API/C/WKContextConfigurationRef.cpp:
(WKContextConfigurationCopyOverrideLanguages): Deleted.
(WKContextConfigurationSetOverrideLanguages): Deleted.
* Source/WebKit/UIProcess/API/C/WKContextConfigurationRef.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivateForTesting.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewTesting.mm:
(+[WKWebView _setOverrideLanguagesForTesting:]):
* Source/WebKit/UIProcess/OverrideLanguages.cpp: Added.
(WebKit::overrideLanguagesStorage):
(WebKit::setOverrideLanguages):
(WebKit::overrideLanguages):
* Source/WebKit/UIProcess/OverrideLanguages.h: Copied from Source/WebKit/UIProcess/WebContextMenuProxy.h.
* Source/WebKit/UIProcess/WebContextMenuProxy.h:
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::setOverrideLanguages):
(WebKit::WebProcessPool::initializeNewWebProcess):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::getLaunchOptions):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::generateContextConfiguration const):
* Tools/WebKitTestRunner/cocoa/TestControllerCocoa.mm:
(WTR::TestController::cocoaResetStateToConsistentValues):

Canonical link: https://commits.webkit.org/252335@main
